### ISSUE TITLE
Add binary parsing primitives?

### ISSUE THREAD

**Issue by ianmackenzie** [mpizenberg/elm-js-typed-array#6]
One of the things I'd be most excited about would be the ability to parse binary formats like Protocol Buffers, MessagePack, BSON etc. High-level parsing functionality (e.g. a binary data version of `elm-tools/parser`) would presumably live in a separate package, but I thought it might make sense to add some low-level functions to this package to read individual primitive values. I was thinking of functions that tried to read a value of a specific type from the start of a `JsArrayBuffer` and return a tuple of ( value, rest of buffer) on success. Reading would fail if there wasn't enough data left in the buffer or if the data was improperly formatted (bad UTF-8 string, etc.). Off the top of my head, something like:

```elm
module JsArrayBuffer

type ReadError
    = EndOfBuffer
    | BadFormat

readUint8 : JsArrayBuffer -> Result ReadError ( Int, JsArrayBuffer )

type Endianness
    = LittleEndian
    | BigEndian

readInt32 : Endianness -> JsArrayBuffer -> Result ReadError ( Int, JsArrayBuffer )

readFloat64 : JsArrayBuffer -> Result ReadError ( Float, JsArrayBuffer )

-- Read a null-terminated ASCII string
readCString : JsArrayBuffer -> Result ReadError ( String, JsArrayBuffer )

-- Read a specific number of bytes in UTF-8 format
readUtf8Bytes : Int -> JsArrayBuffer -> Result ReadError ( String, JsArrayBuffer )

-- Read a specific number of 64-bit floats into a new JsFloat64Array
readFloat64Array : Int -> JsArrayBuffer -> Result ReadError ( JsFloat64Array, JsArrayBuffer )
```

**Comment by mpizenberg** [comment]
Good Idea, that's definitely something doable, but not directly with [`ArrayBuffer`][ArrayBuffer]. An `ArrayBuffer` do not have any `offset` property, meaning the new one returned would be a full copy. However, the good news is that this is exactly the kind of things that can be done by [`DataView`][DataView]. `DataView` isn't really a typed array but is also an [`ArrayBufferView`][ArrayBufferView] like typed arrays, so it makes sense to add it to the package.

In this [introduction to typed arrays][intro-typed-array], there are few examples of how to "read" a buffer to get some values. The way to do so is a bit tedious though since they are manually specifying the address to read in the buffer. So an API to sequentially read a `DataView` of an array buffer could make sense.

One important thing to notice is that `TypedArray` has no control over endianness and uses the one chosen by platform. Only `DataView` can control endianness, but does not have an array-like interpretation, which would come in handy if you'd like to extract an array of numbers out of your protobuf.

Do you have a good reference for the different things that could be useful to read from protobuf?


[ArrayBuffer]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer
[DataView]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView
[ArrayBufferView]: https://developer.mozilla.org/en-US/docs/Web/API/ArrayBufferView
[intro-typed-array]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Typed_arrays

**Comment by mpizenberg** [comment]
Putting this reference here for later: https://www.html5rocks.com/en/tutorials/webgl/typed_arrays/

**Comment by ianmackenzie** [comment]
It's not much, but here's what I could find about the low-level binary read/write API exposed by protobuf: https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.io.coded_stream

**Comment by mpizenberg** [comment]
Feel like I'm gonna need a protobuf crash course ahah. So I am adding these references here for later also:
* [Protobuf documentation home page](https://developers.google.com/protocol-buffers/)
* [Protobuf Github repo](https://github.com/google/protobuf)

Also some JavaScript libraries whose implementations might be inspirationnal:
* [`DataStream.js`][datastream]: improve reading of data from array buffers
* [`text-encoding`][text-encoding]: ease encoding / decoding of strings

[datastream]: https://github.com/kig/DataStream.js
[text-encoding]: https://github.com/inexorabletash/text-encoding
